aspnet/fundamentals/routing.rst

```diff
@@ -1,10 +1,272 @@
+
+1. ``monthly/{blog}/{year}/{month=1}``
+2. ``annual/{blog}/{year}``
+
+.. note:: There is a special case route for filenames, such that you can define a route value like ``files/{filename}.{ext?}``. When both ``filename`` and ``ext`` exist, both values will be populated. However, if only ``filename`` exists in the URL, the trailing period ``.`` is also optional. Thus, these would both match: ``/files/foo.txt`` and ``/files/foo``.
+
+.. tip:: Enable :doc:`logging` to see how the built in routing implementations, such as ``TemplateRoute``, match requests.
+
+Route Builder Extensions
+------------------------
+Several `extension methods on RouteBuilder <https://github.com/aspnet/Routing/blob/1.0.0-rc1/src/Microsoft.AspNet.Routing/RouteBuilderExtensions.cs>`_ are available for convenience. The most common of these is ``MapRoute``, which allows the specification of a route given a name and template, and optionally default values, constraints, and/or data tokens. When using these extensions, you must have specified the ``DefaultHandler`` and ``ServiceProvider`` properties of the ``RouteBuilder`` instance to which you're adding the route. These extensions all make use of the ``TemplateRoute`` type to add new route templates to whichever ``IRouter`` is configured to as the ``DefaultHandler``.
+
+.. note:: ``MapRoute`` doesn't take an ``IRouter`` parameter - it only adds routes to whatever the ``DefaultHandler`` is. This means that if the default is itself a route with named route value placeholders, only these placeholders can be used by routes defined using ``MapRoute``.
+
+If you want ``MapRoute`` to work for any route template, you'll need to ensure the ``DefaultHandler`` is set to use an ``IRouter`` instance that can accept any value. The following ``DelegateRouter`` accepts any values and displays them:
+
+.. literalinclude:: routing/sample/RoutingSample/DelegateRouter.cs
+  :linenos:
+  :language: c#
+
+You configure this ``IRouter`` implementation as the ``DefaultHandler`` for a ``RouteBuilder`` in the ``Configure`` method in ``Startup``:
+
+.. literalinclude:: routing/sample/RoutingSample/Startup.cs
+  :linenos:
+  :language: c#
+  :lines: 27-49
+  :dedent: 12
+  :emphasize-lines: 9,16
+
+
+Link Generation
+---------------
+
+Routing is also used to generate URLs based on route definitions. This is used by helpers to generate links to known actions on controllers in ASP.NET MVC, but can also be used independent of MVC. Given a set of route values, and optionally a route name, you can produce a ``VirtualPathContext`` object. Using the ``VirtualPathContext`` object along with a ``RouteCollection``, you can generate a ``VirtualPath``.
```

Comment:
> This is used by helpers to generate links to known actions on controllers in ASP.NET MVC

This is used by helpers to generate links to known actions on MVC ``:doc:`controllers </mvc/controllers/index>` ``